website/blog/2025-07-01-mlflow-prompt-evaluate/index.md

```diff
@@ -0,0 +1,435 @@
+
+```python
+correct_format = mlflow.metrics.genai.make_genai_metric(
+    name="correct_format",
+    definition=(
+        """The answer is a list of dicts where keys are `question` and `answer`."""
+    ),
+    grading_prompt=(
+        """If formatted correctly, return 1. Otherwise, return 0."""
+    ),
+    model="openai:/gpt-4o-mini",
+    greater_is_better=True,
+)
+```
+
+#### 2. MLflow Evaluate
+
+In this section, we pass the defined metric _correct_format_ to `mlflow.evaluate()`, which makes for every image selected earlier. The JSONs created from the OCR extraction using `get_json()` can used as the ground truth
+
+```python
+def batch_completion(df: pd.DataFrame) -> list[str]:
+    return [get_completion(image) for image in df["inputs"]]
+
+eval_result = mlflow.evaluate(
+    model=batch_completion,
+    data=pd.DataFrame({"inputs": images, "truth": jsons}),
+    targets="truth",
+    model_type="text",
+    extra_metrics=[correct_format],
+)
+```
+
+[WIP] - Add a screenshot of the MLflow UI + explanation
+
+### Applying Prompt Registry and Evaluate to the OCR use case
+
+Let's assume you are developing an OCR application, with different roles contributing to the project development. Here's how a Prompt engineer and a ML engineer benefit from these MLflow capabilities
+
+#### Prompt Engineer: Improve the Prompt
+
+As a Prompt Engineer, you'll be iteratively modifying the prompts and `mlflow.genai.register_prompt()` lets you version different prompts/ system messages that you craft, with the provision to set a template text, and include prompt metadata like author, project, etc. It is also possible to compare different versions of the prompt, and also search for specific prompts from the past.
+
+Here's an example of a prompt template, instructing the LLM to generate
+
+````python
+new_template = """\
+You are an expert at key information extraction and OCR.
+
+Format as a list of dictionaries as shown below. They keys should only be `question` and `answer`.
+
+```\
+[
+    {
+        "question": "question field",
+        "answer": "answer to question field"
+
+    },
+...
+]
+```\
+
+Question refers to a field in the form that takes in information. Answer refers to the information
+that is filled in the field.
+
+Follow these rules:
+- Only use the information present in the text.
+{{ additional_rules }}
+"""
+````
+
+```python
+# Register a new version of an existing prompt
+updated_prompt = mlflow.register_prompt(
+    name="ocr-question-answer",
+    template=new_template,
+    version_metadata={
+        "author": "author@example.com",
+    },
+)
+
+```
```

Comment:
The point of this is the prompt registry decouples prompt engineering from the more-technical aspects of GenAI development. Now non-technical SMEs can work on AI applications

Also because it's registered to a registry, you can easily version and update asynchronously in a CD fashion. Look up the value proposition of model aliasing - it's the same concept for the deployment aspect